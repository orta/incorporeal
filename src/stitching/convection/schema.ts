import { readFileSync } from "fs"
import { makeRemoteExecutableSchema, RenameTypes, transformSchema } from "graphql-tools"
import { createConvectionLink } from "./link"

export const executableConvectionSchema = () => {
  const convectionLink = createConvectionLink()
  const convectionTypeDefs = readFileSync("src/schemas/convection.graphql", "utf8")

  // Setup the default Schema
  const schema = makeRemoteExecutableSchema({
    schema: convectionTypeDefs,
    link: convectionLink,
  })

  // Remap the names of certain types from Convection to fit in the larger
  // metaphysics ecosystem.
  const remap = {
    Submission: "ConsignmentSubmission",
    Category: "ConsignmentSubmissionCategoryAggregation",
    Asset: "ConsignmentSubmissionCategoryAsset",
    State: "ConsignmentSubmissionStateAggregation",
    SubmissionConnection: "ConsignmentSubmissionConnection",
  }

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      const newName = remap[name] || name
      return newName
    }),
  ])
}
