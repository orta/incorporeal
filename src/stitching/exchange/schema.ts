import { readFileSync } from "fs"
import { makeRemoteExecutableSchema, RenameRootFields, RenameTypes, transformSchema } from "graphql-tools"
import { createExchangeLink } from "./link"

export const executableExchangeSchema = () => {
  const exchangeSDL = readFileSync("src/schemas/exchange.graphql", "utf8")
  const exchangeLink = createExchangeLink()

  const schema = makeRemoteExecutableSchema({
    schema: exchangeSDL,
    link: exchangeLink,
  })

  // Return the new modified schema
  return transformSchema(schema, transformsForExchange)
  // Note that changes in this will need to be
}

export const transformsForExchange = [
  // Apply a prefix to all the typenames
  new RenameTypes(name => {
    return `Ecommerce${name}`
  }),
  // Rename all the root fields to be camelCased
  new RenameRootFields((_operation, name) => `ecommerce${name.charAt(0).toUpperCase() + name.slice(1)}`),
]
