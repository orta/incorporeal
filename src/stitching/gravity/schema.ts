import { readFileSync } from "fs"
import {
  FilterTypes,
  makeRemoteExecutableSchema,
  RenameRootFields,
  RenameTypes,
  transformSchema,
} from "graphql-tools"
import { createGravityLink } from "./link"

export const executableGravitySchema = () => {
  const gravityTypeDefs = readFileSync("src/schemas/gravity.graphql", "utf8")

  const gravityLink = createGravityLink()
  const schema = makeRemoteExecutableSchema({
    schema: gravityTypeDefs,
    link: gravityLink,
  })

  // Types which come from Gravity which MP already has copies of.
  // In the future, these could get merged into the MP types.
  const removeTypes = ["Artist", "Artwork"]

  // Return the new modified schema
  return transformSchema(schema, [
    // Remove types which Metaphysics handles better
    new FilterTypes(type => {
      return !removeTypes.includes(type.name)
    }),
    // So, we cannot remove all of the types from a schema is a lesson I have
    // learned in creating these transformations. This means that there has to
    // be at least one type still inside the Schema (excluding the Mutation or
    // the Query)
    // When Partner was removed, we'd see this error
    // https://github.com/graphcool/graphql-import/issues/73
    // but I don't think we're exhibiting the same bug.
    new RenameTypes(name => {
      if (name === "Partner") {
        return "DoNotUseThisPartner"
      }
      return name
    }),
    // We have the same restrictions for root, so let's prefix
    // for now
    new RenameRootFields((type, name, _field) => {
      if (type === "Query") {
        return `_unused_gravity_${name}`
      } else {
        return name
      }
    }),
  ])
}
