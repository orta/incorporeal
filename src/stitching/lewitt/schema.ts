import { readFileSync } from "fs"
import { makeRemoteExecutableSchema, RenameTypes, transformSchema } from "graphql-tools"
import { createLewittLink } from "./link"

export const executableLewittSchema = async () => {
  const lewittSDL = readFileSync("src/schemas/lewitt.graphql", "utf8")
  const lewittLink = createLewittLink()

  const schema = await makeRemoteExecutableSchema({
    schema: lewittSDL,
    link: lewittLink,
  })

  // Remap the names of certain types from Lewitt to fit in the larger
  // metaphysics ecosystem.
  const remap = {
    Currencies: "PartnerProductInvoiceCurrencies",
    Invoice: "PartnerProductInvoice",
    CreateInvoiceInput: "CreatePartnerProductInvoiceInput",
    ArtworkGroup: "PartnerProductInvoiceArtworkGroup",
    LineItem: "PartnerProductInvoiceLineItem",
    MerchantAccount: "PartnerProductMerchantAccount",
    Json: "PartnerProductJson",
  }

  // Return the new modified schema
  return transformSchema(schema, [
    new RenameTypes(name => {
      const newName = remap[name]
      if (!newName) {
        throw new Error(`All types inside Lewitt should be mapped.\n Missing ${name}`)
      }
      return newName
    }),
  ])
}
