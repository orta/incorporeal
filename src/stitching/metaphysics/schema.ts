import { readFileSync } from "fs"
import { makeRemoteExecutableSchema } from "graphql-tools"
import { createMetaphysicsLink } from "./link"

export const executableMetaphysicsSchema = () => {
  const mpTypeDefs = readFileSync("src/schemas/metaphysics.graphql", "utf8")
  const mpLink = createMetaphysicsLink()

  return makeRemoteExecutableSchema({
    schema: mpTypeDefs,
    link: mpLink,
  })
}
