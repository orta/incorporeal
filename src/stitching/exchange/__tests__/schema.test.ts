import { printSchema } from "graphql"
import {
  getMutationFieldsFromSchema,
  getRootFieldsFromSchema,
  getTypesFromSchema,
} from "lib/stitching/lib/getTypesFromSchema"
import { executableExchangeSchema } from "../schema"

it("Does not include generic type names", async () => {
  const exchangeSchema = await executableExchangeSchema()
  const exchangeTypes = await getTypesFromSchema(exchangeSchema)

  expect(exchangeTypes).not.toContain("Order")
  expect(exchangeTypes).not.toContain("LineItem")
  expect(exchangeTypes).not.toContain("DateTime")

  expect(exchangeTypes).toContain("EcommerceOrder")
})

it("has all our root fields", async () => {
  const exchangeSchema = await executableExchangeSchema()
  const rootFields = await getRootFieldsFromSchema(exchangeSchema)

  expect(rootFields).not.toContain("order")
  expect(rootFields).toContain("ecommerceOrder")
})

it("Includes prefixed mutations", async () => {
  const exchangeSchema = await executableExchangeSchema()
  const mutations = await getMutationFieldsFromSchema(exchangeSchema)

  expect(mutations).toContain("ecommerceCreateOrderWithArtwork")
  expect(mutations).toContain("ecommerceApproveOrder")
})

it("creates an SDL", async () => {
  const exchangeSchema = await executableExchangeSchema()
  expect(printSchema(exchangeSchema, { commentDescriptions: true })).toMatchSnapshot()
})
