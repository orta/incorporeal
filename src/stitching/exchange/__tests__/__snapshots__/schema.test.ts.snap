// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates an SDL 1`] = `
"# An generic error type for mutations
type EcommerceApplicationError {
  # Code of this error
  code: String!

  # What caused the error
  data: String

  # Type of this error
  type: String!
}

# Autogenerated input type of ApproveOrder
input EcommerceApproveOrderInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  id: ID!
}

# Autogenerated return type of ApproveOrder
type EcommerceApproveOrderPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

# Autogenerated input type of CreateOrderWithArtwork
input EcommerceCreateOrderWithArtworkInput {
  # Artwork Id
  artworkId: String!

  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # EditionSet Id
  editionSetId: String

  # Number of items in the line item
  quantity: Int
}

# Autogenerated return type of CreateOrderWithArtwork
type EcommerceCreateOrderWithArtworkPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

scalar EcommerceDate

scalar EcommerceDateTime

# Autogenerated input type of FulfillAtOnce
input EcommerceFulfillAtOnceInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  fulfillment: EcommerceFulfillmentAttributes!
  id: ID!
}

# Autogenerated return type of FulfillAtOnce
type EcommerceFulfillAtOncePayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

# A Fulfillment for an order
type EcommerceFulfillment {
  courier: String!
  createdAt: EcommerceDateTime!
  estimatedDelivery: EcommerceDate
  id: ID!
  notes: String
  trackingId: String
  updatedAt: EcommerceDateTime!
}

# Attributes of a Fulfillment
input EcommerceFulfillmentAttributes {
  courier: String!
  estimatedDelivery: EcommerceDate
  notes: String
  trackingId: String
}

# The connection type for Fulfillment.
type EcommerceFulfillmentConnection {
  # A list of edges.
  edges: [EcommerceFulfillmentEdge]

  # A list of nodes.
  nodes: [EcommerceFulfillment]

  # Information to aid in pagination.
  pageInfo: EcommercePageInfo!
}

# An edge in a connection.
type EcommerceFulfillmentEdge {
  # A cursor for use in pagination.
  cursor: String!

  # The item at the end of the edge.
  node: EcommerceFulfillment
}

# A Line Item
type EcommerceLineItem {
  artworkId: String!
  createdAt: EcommerceDateTime!
  editionSetId: String
  fulfillments(
    # Returns the elements in the list that come after the specified cursor.
    after: String

    # Returns the elements in the list that come before the specified cursor.
    before: String

    # Returns the first _n_ elements from the list.
    first: Int

    # Returns the last _n_ elements from the list.
    last: Int
  ): EcommerceFulfillmentConnection
  id: ID!
  priceCents: Int!
  quantity: Int!
  updatedAt: EcommerceDateTime!
}

# The connection type for LineItem.
type EcommerceLineItemConnection {
  # A list of edges.
  edges: [EcommerceLineItemEdge]

  # A list of nodes.
  nodes: [EcommerceLineItem]

  # Information to aid in pagination.
  pageInfo: EcommercePageInfo!
}

# An edge in a connection.
type EcommerceLineItemEdge {
  # A cursor for use in pagination.
  cursor: String!

  # The item at the end of the edge.
  node: EcommerceLineItem
}

# An Order
type EcommerceOrder {
  buyer: EcommerceOrderPartyUnion!
  buyerPhoneNumber: String
  buyerTotalCents: Int
  code: String!
  commissionFeeCents: Int
  createdAt: EcommerceDateTime!
  creditCardId: String
  currencyCode: String!
  id: ID!
  itemsTotalCents: Int!
  lastApprovedAt: EcommerceDateTime
  lastSubmittedAt: EcommerceDateTime
  lineItems(
    # Returns the elements in the list that come after the specified cursor.
    after: String

    # Returns the elements in the list that come before the specified cursor.
    before: String

    # Returns the first _n_ elements from the list.
    first: Int

    # Returns the last _n_ elements from the list.
    last: Int
  ): EcommerceLineItemConnection
  requestedFulfillment: EcommerceRequestedFulfillmentUnion
  seller: EcommerceOrderPartyUnion!
  sellerTotalCents: Int
  shippingTotalCents: Int
  state: EcommerceOrderStateEnum!
  stateExpiresAt: EcommerceDateTime
  stateUpdatedAt: EcommerceDateTime
  taxTotalCents: Int
  transactionFeeCents: Int
  updatedAt: EcommerceDateTime!
}

# Fields to sort by
enum EcommerceOrderConnectionSortEnum {
  # Sort by the timestamp the state of order was last updated in ascending order
  STATE_UPDATED_AT_ASC

  # Sort by the timestamp the state of order was last updated in descending order
  STATE_UPDATED_AT_DESC

  # Sort by the timestamp the order was last updated in ascending order
  UPDATED_AT_ASC

  # Sort by the timestamp the order was last updated in descending order
  UPDATED_AT_DESC
}

# The connection type for Order.
type EcommerceOrderConnectionWithTotalCount {
  # A list of edges.
  edges: [EcommerceOrderEdge]

  # A list of nodes.
  nodes: [EcommerceOrder]

  # Information to aid in pagination.
  pageInfo: EcommercePageInfo!
  totalCount: Int!
}

# An edge in a connection.
type EcommerceOrderEdge {
  # A cursor for use in pagination.
  cursor: String!

  # The item at the end of the edge.
  node: EcommerceOrder
}

enum EcommerceOrderFulfillmentTypeEnum {
  # fulfillment type is: pickup
  PICKUP

  # fulfillment type is: ship
  SHIP
}

# Represents either a resolved Order or a potential failure
union EcommerceOrderOrFailureUnion = EcommerceOrderWithMutationFailure | EcommerceOrderWithMutationSuccess

# Represents either a partner or a user
union EcommerceOrderPartyUnion = EcommercePartner | EcommerceUser

enum EcommerceOrderStateEnum {
  # order is abandoned by buyer and never submitted
  ABANDONED

  # order is approved by seller
  APPROVED

  # order is canceled
  CANCELED

  # order is fulfilled by seller
  FULFILLED

  # order is still pending submission by buyer
  PENDING

  # order is submitted by buyer
  SUBMITTED
}

# An error response for changes to an order
type EcommerceOrderWithMutationFailure {
  error: EcommerceApplicationError!
}

# A successfully returned order type
type EcommerceOrderWithMutationSuccess {
  order: EcommerceOrder!
}

# Information about pagination in a connection.
type EcommercePageInfo {
  # When paginating forwards, the cursor to continue.
  endCursor: String

  # When paginating forwards, are there more items?
  hasNextPage: Boolean!

  # When paginating backwards, are there more items?
  hasPreviousPage: Boolean!

  # When paginating backwards, the cursor to continue.
  startCursor: String
}

type EcommercePartner {
  id: String!
}

type EcommercePickup {
  fulfillmentType: String!
}

# Autogenerated input type of RejectOrder
input EcommerceRejectOrderInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  id: ID!
}

# Autogenerated return type of RejectOrder
type EcommerceRejectOrderPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

# Represents either a shipping information or pickup
union EcommerceRequestedFulfillmentUnion = EcommercePickup | EcommerceShip

# Autogenerated input type of SetPayment
input EcommerceSetPaymentInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  creditCardId: String!
  id: ID!
}

# Autogenerated return type of SetPayment
type EcommerceSetPaymentPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

# Autogenerated input type of SetShipping
input EcommerceSetShippingInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  fulfillmentType: EcommerceOrderFulfillmentTypeEnum!
  id: ID!
  shipping: EcommerceShippingAttributes
}

# Autogenerated return type of SetShipping
type EcommerceSetShippingPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

type EcommerceShip {
  addressLine1: String
  addressLine2: String
  city: String
  country: String
  name: String
  phoneNumber: String
  postalCode: String
  region: String
}

# Shipping information
input EcommerceShippingAttributes {
  addressLine1: String
  addressLine2: String
  city: String
  country: String
  name: String
  phoneNumber: String!
  postalCode: String
  region: String
}

# Autogenerated input type of SubmitOrder
input EcommerceSubmitOrderInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  id: ID!
}

# Autogenerated return type of SubmitOrder
type EcommerceSubmitOrderPayload {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String

  # A union of success/failure
  orderOrError: EcommerceOrderOrFailureUnion!
}

type EcommerceUser {
  id: String!
}

type Mutation {
  ecommerceApproveOrder(input: EcommerceApproveOrderInput!): EcommerceApproveOrderPayload
  ecommerceCreateOrderWithArtwork(input: EcommerceCreateOrderWithArtworkInput!): EcommerceCreateOrderWithArtworkPayload

  # Fulfill an order with one Fulfillment, it sets this fulfillment to each line item in order
  ecommerceFulfillAtOnce(input: EcommerceFulfillAtOnceInput!): EcommerceFulfillAtOncePayload
  ecommerceRejectOrder(input: EcommerceRejectOrderInput!): EcommerceRejectOrderPayload
  ecommerceSetPayment(input: EcommerceSetPaymentInput!): EcommerceSetPaymentPayload
  ecommerceSetShipping(input: EcommerceSetShippingInput!): EcommerceSetShippingPayload
  ecommerceSubmitOrder(input: EcommerceSubmitOrderInput!): EcommerceSubmitOrderPayload
}

type Query {
  # Find an order by ID
  ecommerceOrder(id: ID!): EcommerceOrder

  # Find list of orders
  ecommerceOrders(
    # Returns the elements in the list that come after the specified cursor.
    after: String

    # Returns the elements in the list that come before the specified cursor.
    before: String
    buyerId: String
    buyerType: String

    # Returns the first _n_ elements from the list.
    first: Int

    # Returns the last _n_ elements from the list.
    last: Int
    sellerId: String
    sellerType: String
    sort: EcommerceOrderConnectionSortEnum
    state: EcommerceOrderStateEnum
  ): EcommerceOrderConnectionWithTotalCount
}
"
`;
